### PR TITLE
AVRO-3603 dotnet reflect reader and writer refactoring

### PR DESCRIPTION
**Pull request by KhrystynaPopadyuk**
### Jira

- [ ] My PR addresses the following [Avro Jira](https://issues.apache.org/jira/browse/AVRO-3603) issues and references them in the PR title. 

### Tests
My PR do not add or update functionality, there is no new tests

### Commits

- [ ] My commits all reference Jira issues in their subject lines. 

### Documentation

Readme file is updated
https://github.com/KhrystynaPopadyuk/avro/tree/avro-3603-dotnet-reflect-refactoring/lang/csharp/src/apache/main/Reflect

### CODE REVIEW

File: lang/csharp/src/apache/main/Reflect/ReflectDefaultWriter.cs [hunk 1]
```diff
@@ -188,9 +181,9 @@
                     return obj is string;
                 case Schema.Type.Error:
                 case Schema.Type.Record:
-                    return _classCache.GetClass(sc as RecordSchema).GetClassType() == obj.GetType();
+                    return _reflectCache.GetClass(sc as RecordSchema).GetClassType() == obj.GetType();
```
[the remark this answers is not in the document]
Comment (KhrystynaPopadyuk) [reply lang/csharp/src/apache/main/Reflect/ReflectDefaultWriter.cs:184]: This is existing logic.

File: lang/csharp/src/apache/main/Reflect/ReflectDefaultWriter.cs [hunk 2]
```diff
@@ -188,9 +181,9 @@
...
                 case Schema.Type.Enumeration:
-                    return EnumCache.GetEnumeration(sc as EnumSchema) == obj.GetType();
+                    return _reflectCache.GetEnumeration(sc as EnumSchema) == obj.GetType();
```
[the remark this answers is not in the document]
Comment (KhrystynaPopadyuk) [reply lang/csharp/src/apache/main/Reflect/ReflectDefaultWriter.cs:186]: This is existing logic. Should not be updated in scope of AVRO-3603

File: lang/csharp/src/apache/main/Reflect/Reflection/ReflectCache.cs [hunk 3]
```diff
@@ -288,17 +256,53 @@
...
+            {
+                return;
+            }
+
+            if (!dotnetClass.IsClass)
+            {
+                throw new AvroException($"Type {dotnetClass.Name} is not a class");
+            }
+
+            _nameClassMap.TryAdd(schema.Fullname, new DotnetClass(dotnetClass, schema, this));
```
[the remark this answers is not in the document]
Comment (KhrystynaPopadyuk) [reply lang/csharp/src/apache/main/Reflect/Reflection/ReflectCache.cs:305]: This is existing logic. Should not be updated in scope of AVRO-3603

File: lang/csharp/src/apache/main/Reflect/ReflectDefaultReader.cs [hunk 4]
```diff
@@ -37,12 +39,7 @@
         /// </summary>
         public Type MapType { get => _mapType; set => _mapType = value; }
 
-        private ClassCache _classCache = new ClassCache();
-
-        /// <summary>
-        /// Class cache
-        /// </summary>
-        public ClassCache ClassCache { get => _classCache; }
+        private IReflectCache _reflectCache;
```
[the remark this answers is not in the document]
Comment (KhrystynaPopadyuk) [reply lang/csharp/src/apache/main/Reflect/ReflectDefaultReader.cs]: Done

File: lang/csharp/src/apache/main/Reflect/ReflectDefaultWriter.cs [hunk 5]
```diff
@@ -28,28 +30,19 @@
     /// </summary>
     public class ReflectDefaultWriter : SpecificDefaultWriter
     {
-        private ClassCache _classCache = new ClassCache();
-
-        /// <summary>
-        /// Class cache
-        /// </summary>
-        public ClassCache ClassCache { get => _classCache; }
+        private IReflectCache _reflectCache;
```
[the remark this answers is not in the document]
Comment (KhrystynaPopadyuk) [reply lang/csharp/src/apache/main/Reflect/ReflectDefaultWriter.cs]: Done

File: lang/csharp/src/apache/main/Reflect/Reflection/ReflectCache.cs [hunk 6]
```diff
@@ -19,63 +19,31 @@
...
 {
-    /// <summary>
-    /// Class holds a cache of C# classes and their properties. The key for the cache is the schema full name.
-    /// </summary>
-    public class ClassCache
+    internal class ReflectCache : IReflectCache
     {
-        private static ConcurrentBag<IAvroFieldConverter> _defaultConverters = new ConcurrentBag<IAvroFieldConverter>();
-
+        private ConcurrentDictionary<string, Type> _nameEnumMap = new ConcurrentDictionary<string, Type>();
```
[the remark this answers is not in the document]
Comment (KhrystynaPopadyuk) [reply lang/csharp/src/apache/main/Reflect/Reflection/ReflectCache.cs]: Done

**Comment by KhrystynaPopadyuk** [comment]
Hi @KalleOlaviNiemitalo and @RyanSkraba ,

Please review my new PR. This is refactoring to reduce static class and fields and introduce DI. 
This PR will allow further refactoring in future.

Thanks.

File: lang/csharp/src/apache/main/Reflect/Converter/DateTimeOffsetToLongConverter.cs [hunk 7]
```diff
@@ -17,8 +17,9 @@
  */
 
 using System;
+using Avro.Reflect.Interface;
 
-namespace Avro.Reflect
+namespace Avro.Reflect.Conver
```
Comment (martin-g) [inline lang/csharp/src/apache/main/Reflect/Converter/DateTimeOffsetToLongConverter.cs:22]: missing `er` ?!

File: lang/csharp/src/apache/main/Reflect/EnumCache.cs [hunk 8]
```diff
@@ -1,57 +0,0 @@
...
-using System;
-using System.Collections.Concurrent;
-using Avro;
-
-namespace Avro.Reflect
-{
-    /// <summary>
-    /// Cache of enum types. Cache key is the schema fullname.
-    /// </summary>
-    public static class EnumCache
```
Comment (martin-g) [inline lang/csharp/src/apache/main/Reflect/EnumCache.cs:27]: Is it possible to keep the class for a while as deprecated ?

File: lang/csharp/src/apache/main/Reflect/Reflection/DotnetClass.cs [hunk 9]
```diff
@@ -20,8 +20,9 @@
 using System.Reflection;
 using System.Collections.Concurrent;
 using Avro;
+using Avro.Reflect.Interface;
 
-namespace Avro.Reflect
+namespace Avro.Reflect.Reflection
```
Comment (martin-g) [inline lang/csharp/src/apache/main/Reflect/Reflection/DotnetClass.cs:25]: Why there are two `reflect***` packages/namespaces here ?

**Review by martin-g — COMMENTED** [review]
IMO this PR introduces too much API breaks that will lead to a lot of complains by other users.

**Comment by KhrystynaPopadyuk** [comment]
Hi @KalleOlaviNiemitalo  and @martin-g ,

Thank you for review and feedback.

I believe that at the end of day such refactoring is important and would be beneficial for both contributors (provide easier maintenance and extensibility) and end users (make it flexible, allow inject custom logic and override default behavior).
There is no goal to change serialization and deserialization. All changes are around reflection.

From your comments I understand that there are two main concerns: breaking changes and new dependency.

I agree with @KalleOlaviNiemitalo that we can easy resolve new dependency issue - do not add DI, create new project or leave it with end user.

About breaking changes. After second thought it looks like there is way to do refactoring without introducing breaking changes. The example and final state can be reviewed at https://github.com/apache/avro/pull/1823. As you can see in case if there is no breaking changes there are much more updates, that PR is massive.

I do not think that it is safe to deliver such massive PR. Also it is difficult to review them. Likely we can break down it to small pieces, implement, review and deliver it one by one. First such update are created and ready for review: https://github.com/apache/avro/pull/1824. 

Thanks one more time for your time.
Khrystyna

**Pull request closed without merging**